libc/watcom/syscall/fmemalloc.c

```diff
@@ -0,0 +1,42 @@
+/****************************************************************************
+*
+*                            Open Watcom Project
+*
+* Copyright (c) 2002-2019 The Open Watcom Contributors. All Rights Reserved.
+*    Portions Copyright (c) 1983-2002 Sybase, Inc. All Rights Reserved.
+*
+*  ========================================================================
+*
+*    This file contains Original Code and/or Modifications of Original
+*    Code as defined in and that are subject to the Sybase Open Watcom
+*    Public License version 1.0 (the 'License'). You may not use this file
+*    except in compliance with the License. BY USING THIS FILE YOU AGREE TO
+*    ALL TERMS AND CONDITIONS OF THE LICENSE. A copy of the License is
+*    provided with the Original Code and Modifications, and is also
+*    available at www.sybase.com/developer/opensource.
+*
+*    The Original Code and all software distributed under the License are
+*    distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
+*    EXPRESS OR IMPLIED, AND SYBASE AND ALL CONTRIBUTORS HEREBY DISCLAIM
+*    ALL SUCH WARRANTIES, INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF
+*    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR
+*    NON-INFRINGEMENT. Please see the License for the specific language
+*    governing rights and limitations under the License.
+*
+*  ========================================================================
+*
+* Description:  ELKS _fmemalloc() system call.
+*
+****************************************************************************/
+
+
+#include <malloc.h>
+#include "watcom/syselks.h"
+
+int _fmemalloc( int __paras, unsigned short *__pseg )
+{
+    sys_set_ptr_seg(__pseg);
```

Comment:
There is a bug here calling `sys_set_ptr_seg`, and the line should be deleted. The reason is a bit complicated: in the `_sbrk` code, the argument is `void **__pnewbrk`, which is a pointer to a pointer, unlike here, where we only have a pointer to an unsigned short.

So there are two pointer indirections. The first indirection is used by the kernel, and the `sys_setseg` call sets DS to the segment to be used, and the offset is passed in `(unsigned)__pnewbrk` (see sbrk.c). Thus the kernel gets the DS value from `current->t_regs.ds` and the offset from the (always near) `__pnewbrk` in the kernel.

The problem is that in large model, we actually need a system call that returns a pointer to another pointer. That's the `void *` that the first `*` is pointing to. Now, in large model, that pointer is 4 bytes long, and the offset has been set by the kernel, but the segment portion the kernel knows nothing about. So, the definition of `sys_set_ptr_seg` is:
```
#if defined(__COMPACT__) || defined(__LARGE__)
#define sys_set_ptr_seg(ptr)    \
    (*((unsigned *)(ptr) + 1) = (unsigned long)(ptr) >> 16)
#else
#define sys_set_ptr_seg(ptr)    /* DS already set in return pointer in small/medium */
#endif
```
This bit of magic code sets the segment portion of the second pointer to the segment being passed to the system call (which will almost always be DS or SS).

In our case for `_fmemalloc`, the system call only returns a segment number, not a pointer. Thus, this call writes two bytes of segment in the memory address 2 bytes past the address of the variable the programmer passed to the system call wrapper!

In the small/medium case, pointers are only 2 bytes, so the `sys_set_ptr_seg` does nothing. In all cases, this should not be performed for `_fmemalloc`.

I hope this makes sense!!

Reply:
Hmm, difficult:)
I will try to understand, and the fix the codes but is this mean the small model has not been affected?
If it is so, the root cause of the program not working may be different.

Thank you.